src/routing/approved.rs

```diff
@@ -39,22 +38,83 @@ use bytes::Bytes;
 use ed25519_dalek::Verifier;
 use itertools::Itertools;
 use resource_proof::ResourceProof;
+use sn_data_types::{PublicKey as EndUserPK, Signature as EndUserSig};
 use sn_messaging::{
-    infrastructure::{
-        Error as InfrastructureError, GetSectionResponse, InfrastructureInformation,
-        Message as InfrastructureMessage,
-    },
     node::NodeMessage,
-    MessageType,
+    section_info::{
+        Error as TargetSectionError, GetSectionResponse, Message as SectionInfoMsg, SectionInfo,
+    },
+    DstLocation, EndUser, MessageType, SrcLocation,
+};
+use std::{
+    cmp,
+    collections::{btree_map::Entry, BTreeMap},
+    net::SocketAddr,
+    slice,
 };
-use std::{cmp, net::SocketAddr, slice};
 use tokio::sync::mpsc;
 use xor_name::{Prefix, XorName};
 
 pub(crate) const RESOURCE_PROOF_DATA_SIZE: usize = 64;
 pub(crate) const RESOURCE_PROOF_DIFFICULTY: u8 = 2;
 const KEY_CACHE_SIZE: u8 = 5;
 
+type SocketId = XorName;
+
+struct EndUserRegistry {
+    clients: BTreeMap<SocketAddr, EndUser>,
+    socket_id_mapping: BTreeMap<SocketId, SocketAddr>,
+}
+
+impl EndUserRegistry {
+    pub fn new() -> Self {
+        Self {
+            clients: BTreeMap::default(),
+            socket_id_mapping: BTreeMap::default(),
+        }
+    }
+
+    pub fn get_enduser_by_addr(&self, socketaddr: &SocketAddr) -> Option<EndUser> {
```

Comment:
Minor, but why are we returning by value here but by reference in the next function?